File: plugins/plugin-tee/rust/src/providers/remote_attestation.rs
#![allow(missing_docs)]

use async_trait::async_trait;
use tracing::info;

use crate::client::TeeClient;
use crate::error::{Result, TeeError};
use crate::providers::base::RemoteAttestationProvider;
use crate::types::{RemoteAttestationQuote, TdxQuoteHashAlgorithm};
use crate::utils::{current_timestamp_ms, get_tee_endpoint};

pub struct PhalaRemoteAttestationProvider {
    client: TeeClient,
}

impl PhalaRemoteAttestationProvider {
    pub fn new(tee_mode: &str) -> Result<Self> {
        let endpoint = get_tee_endpoint(tee_mode)?;

        if let Some(ref ep) = endpoint {
            info!("TEE: Connecting to simulator at {}", ep);
        } else {
            info!("TEE: Running in production mode without simulator");
        }

        Ok(Self {
            client: TeeClient::new(endpoint),
        })
    }
}

#[async_trait]
impl RemoteAttestationProvider for PhalaRemoteAttestationProvider {
    async fn generate_attestation(
        &self,
        report_data: &str,
        hash_algorithm: Option<TdxQuoteHashAlgorithm>,
    ) -> Result<RemoteAttestationQuote> {
        let hash_algo = hash_algorithm.map(|a| match a {
            TdxQuoteHashAlgorithm::Sha256 => "sha256",
            TdxQuoteHashAlgorithm::Sha384 => "sha384",
            TdxQuoteHashAlgorithm::Sha512 => "sha512",
            TdxQuoteHashAlgorithm::Raw => "raw",
        });

        let result = self
            .client
            .tdx_quote(report_data, hash_algo)
            .await
            .map_err(|e| TeeError::attestation(e.to_string()))?;

        Ok(RemoteAttestationQuote {
            quote: result.quote,
            timestamp: current_timestamp_ms(),
        })
    }
}

File: plugins/plugin-tee/typescript/src/providers/remoteAttestation.ts
/**
 * Remote Attestation Provider for Phala TEE.
 */

import { type Provider, type IAgentRuntime, type Memory, logger } from "@elizaos/core";
import { TappdClient, type TdxQuoteHashAlgorithms, type TdxQuoteResponse } from "@phala/dstack-sdk";
import type { RemoteAttestationQuote, RemoteAttestationMessage, TdxQuoteHashAlgorithm, TeeProviderResult } from "../types";
import { getTeeEndpoint } from "../utils";
import { RemoteAttestationProvider } from "./base";

/**
 * Phala Network Remote Attestation Provider.
 *
 * Generates TDX attestation quotes for proving TEE execution.
 */
export class PhalaRemoteAttestationProvider extends RemoteAttestationProvider {
  private readonly client: TappdClient;

  constructor(teeMode: string) {
    super();
    const endpoint = getTeeEndpoint(teeMode);

    logger.info(
      endpoint
        ? `TEE: Connecting to simulator at ${endpoint}`
        : "TEE: Running in production mode without simulator"
    );

    this.client = endpoint ? new TappdClient(endpoint) : new TappdClient();
  }

  /**
   * Generate a remote attestation quote.
   *
   * @param reportData - The data to include in the attestation report.
   * @param hashAlgorithm - Optional hash algorithm for the quote.
   * @returns The remote attestation quote.
   */
  async generateAttestation(
    reportData: string,
    hashAlgorithm?: TdxQuoteHashAlgorithm
  ): Promise<RemoteAttestationQuote> {
    try {
      logger.debug(`Generating attestation for: ${reportData.substring(0, 100)}...`);

      const tdxQuote: TdxQuoteResponse = await this.client.tdxQuote(
        reportData,
        hashAlgorithm as TdxQuoteHashAlgorithms | undefined
      );

      const rtmrs = tdxQuote.replayRtmrs();
      logger.debug(
        `RTMR values: rtmr0=${rtmrs[0]}, rtmr1=${rtmrs[1]}, rtmr2=${rtmrs[2]}, rtmr3=${rtmrs[3]}`
      );

      const quote: RemoteAttestationQuote = {
        quote: tdxQuote.quote,
        timestamp: Date.now(),
      };

      logger.info("Remote attestation quote generated successfully");
      return quote;
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      logger.error(`Error generating remote attestation: ${message}`);
      throw new Error(`Failed to generate TDX Quote: ${message}`);
    }
  }
}

/**
 * elizaOS Provider for remote attestation.
 *
 * This provider generates attestation based on the current message context.
 */
export const phalaRemoteAttestationProvider: Provider = {
  name: "phala-remote-attestation",

  get: async (runtime: IAgentRuntime, message: Memory): Promise<TeeProviderResult> => {
    const teeMode = runtime.getSetting("TEE_MODE");
    if (!teeMode) {
      return {
        data: null,
        values: {},
        text: "TEE_MODE is not configured",
      };
    }

    const provider = new PhalaRemoteAttestationProvider(teeMode);
    const agentId = runtime.agentId;

    try {
      const attestationMessage: RemoteAttestationMessage = {
        agentId,
        timestamp: Date.now(),
        message: {
          entityId: message.entityId,
          roomId: message.roomId,
          content: message.content.text ?? "",
        },
      };

      logger.debug(`Generating attestation for message: ${JSON.stringify(attestationMessage)}`);

      const attestation = await provider.generateAttestation(
        JSON.stringify(attestationMessage)
      );

      return {
        data: {
          quote: attestation.quote,
          timestamp: attestation.timestamp.toString(),
        },
        values: {
          quote: attestation.quote,
          timestamp: attestation.timestamp.toString(),
        },
        text: `Your Agent's remote attestation is: ${JSON.stringify(attestation)}`,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      logger.error(`Error in remote attestation provider: ${message}`);
      throw new Error(`Failed to generate TDX Quote: ${message}`);
    }
  },
};


